PR: https://github.com/odoo/odoo/pull/262274

From: 6f8c2526a00d12de0ab68edb5279b37048841a8d
From: Claire (clbr)
Date: 2026-06-18 08:23:55

Breaking data model changes scores: del:4 + add:0, change matches:
-    is_peppol_edi_format = fields.Boolean(compute='_compute_is_peppol_edi_format')
-    peppol_endpoint = fields.Char(
-    peppol_eas = fields.Selection(
-    available_peppol_eas = fields.Json(compute='_compute_available_peppol_eas')

Total Changes: 815

[REF] base,*: multi ID for UBL + remove company_registry

We previously replaced the scattered, country-specific partner
identifier fields with one JSON field `additional_identifiers`.
This commits continues to clean up after that change, mainly:

* Move the additional_identifiers JSON field into base: some modules
(l10n_xx_hr_payroll) don't depend on `account` yet still have similar
contact's identifiers.
* Drop the company_registry field as now, everything should be more
specific, and stored in the JSON field.

Smaller changes:
* add base helpers selecting a partner's preferred legal-entity, tax and
routing identifier, reused by UBL/CII EDI, SAF-T and Peppol/PDP
* rename peppol_eas/peppol_endpoint to routing_scheme/routing_endpoint
(+ computed routing_identifier) so any e-invoicing network can route
using this fields => less confusing.
* Some countries have non-stored compute/inverse for convient access
(l10n_fr_siret/siren, SG UEN, NO Brønnøysund, ...),
* adapt UBL/CII exports, SAF-T reports and invoice reports after the
removal of the company_registry.
* cleanup of the scheme/endpoint validation so that's it's more coherent. There
was 3 different validation before, of only a subset of schemes.

task-6140750

Part-of: odoo/odoo#262274
Related: odoo/enterprise#115838
Related: odoo/upgrade#10479
Signed-off-by: Christophe Monniez (moc) <moc@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account_edi_ubl_cii/models/__init__.py
+++ b/addons/account_edi_ubl_cii/models/__init__.py
@@ -16,4 +16,5 @@ from . import account_move
 from . import account_move_send
 from . import account_tax
 from . import ir_actions_report
+from . import res_company
 from . import res_partner

--- a/addons/account_edi_ubl_cii/models/account_edi_common.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_common.py
@@ -15,7 +15,7 @@ from odoo.tools.float_utils import float_round
 from odoo.tools.misc import formatLang, html_escape
 from odoo.tools.xml_utils import find_xml_value
 
-from odoo.addons.account.tools.country_groups import EUROPEAN_ECONOMIC_AREA_COUNTRY_CODES
+from odoo.addons.base.models.res_country import EUROPEAN_ECONOMIC_AREA_COUNTRY_CODES
 
 # -------------------------------------------------------------------------
 # UNIT OF MEASURE
@@ -51,78 +51,6 @@ UOM_TO_UNECE_CODE = {
     'uom.product_uom_kwh': 'KWH',
 }
 
-# -------------------------------------------------------------------------
-# ELECTRONIC ADDRESS SCHEME (EAS), see https://docs.peppol.eu/poacc/billing/3.0/codelist/eas/
-# -------------------------------------------------------------------------
-EAS_MAPPING = {
-    'AD': {'9922': 'vat'},
-    'AE': {'0235': 'vat'},
-    'AL': {'9923': 'vat'},
-    'AT': {'9915': 'vat'},
-    'AU': {'0151': 'vat'},
-    'BA': {'9924': 'vat'},
-    'BE': {'0208': 'additional_identifiers', '9925': 'vat'},
-    'BG': {'9926': 'vat'},
-    'CH': {'9927': 'vat', '0183': None},
-    'CY': {'9928': 'vat'},
-    'CZ': {'9929': 'vat'},
-    'DE': {'9930': 'vat', '0246': 'l10n_de_widnr'},
-    'DK': {'0184': 'vat', '0198': 'vat'},
-    'EE': {'9931': 'vat'},
-    'ES': {'9920': 'vat'},
-    'FI': {'0216': None},
-    'FR': {'0225': 'peppol_endpoint', '0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # `peppol_endpoint` used as place holder for custom logic via `_get_peppol_endpoint_value`
-    'SG': {'0195': 'additional_identifiers'},
-    'GB': {'9932': 'vat'},
-    'GR': {'9933': 'vat'},
-    'HR': {'9934': 'vat', '0088': 'company_registry'},
-    'HU': {'9910': 'l10n_hu_eu_vat'},
-    'IE': {'9935': 'vat'},
-    'IS': {'0196': 'vat'},
-    'IT': {'0211': 'vat', '0210': 'l10n_it_codice_fiscale'},
-    'JP': {'0221': 'vat'},
-    'LI': {'9936': 'vat'},
-    'LT': {'9937': 'vat'},
-    'LU': {'9938': 'vat'},
-    'LV': {'0218': 'company_registry', '9939': 'vat'},
-    'MC': {'9940': 'vat'},
-    'ME': {'9941': 'vat'},
-    'MK': {'9942': 'vat'},
-    'MT': {'9943': 'vat'},
-    'MY': {'0230': None},
-    # Do not add the vat for NL, since: "[NL-R-003] For suppliers in the Netherlands, the legal entity identifier
-    # MUST be either a KVK or OIN number (schemeID 0106 or 0190)" in the Bis 3 rules (in PartyLegalEntity/CompanyID).
-    'NG': {'0244': 'vat'},
-    'NL': {'0106': None, '0190': None},
-    'NO': {'0192': 'additional_identifiers'},
-    'NZ': {'0088': 'company_registry'},
-    'PL': {'9945': 'vat'},
-    'PT': {'9946': 'vat'},
-    'RO': {'9947': 'vat'},
-    'RS': {'9948': 'vat'},
-    'SE': {'0007': 'company_registry', '9955': 'vat'},
-    'SI': {'9949': 'vat'},
-    'SK': {'9950': 'vat', '0245': 'company_registry'},
-    'SM': {'9951': 'vat'},
-    'TR': {'9952': 'vat'},
-    'VA': {'9953': 'vat'},
-    # DOM-TOM
-    'BL': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Saint Barthélemy
-    'GF': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # French Guiana
-    'GP': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Guadeloupe
-    'MF': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Saint Martin
-    'MQ': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Martinique
-    'NC': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # New Caledonia
-    'PF': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # French Polynesia
-    'PM': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Saint Pierre and Miquelon
-    'RE': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Réunion
-    'TF': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # French Southern and Antarctic Lands
-    'WF': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Wallis and Futuna
-    'YT': {'0009': 'additional_identifiers', '9957': 'vat', '0002': None},  # Mayotte
-
-    'AX': {'0216': None},  # Åland Islands
-}
-
 # -------------------------------------------------------------------------
 # MAPPING FOR TAX EXEMPTION
 # -------------------------------------------------------------------------
@@ -738,13 +666,13 @@ class AccountEdiCommon(models.AbstractModel):
 
         return attachments
 
-    def _import_partner(self, company_id, name, phone, email, vat, *, peppol_eas=False, peppol_endpoint=False, additional_identifiers=None, postal_address={}, **kwargs):
+    def _import_partner(self, company_id, name, phone, email, vat, *, routing_identifier=False, additional_identifiers=None, postal_address={}, **kwargs):
         """ Retrieve the partner, if no matching partner is found, create it (only if he has a vat and a name) """
         logs = []
-        if peppol_eas and peppol_endpoint:
-            domain = [('peppol_eas', '=', peppol_eas), ('peppol_endpoint', '=', peppol_endpoint)]
-        else:
-            domain = False
+        domain = False
+        if routing_identifier:
+            scheme, _sep, endpoint = routing_identifier.partition(':')
+            domain = [('routing_scheme', '=', scheme), ('routing_endpoint', '=', endpoint)]
         partner = self.env['res.partner'] \
             .with_company(company_id) \
             ._retrieve_partner(name=name, phone=phone, email=email, vat=vat, additional_identifiers=additional_identifiers, domain=domain)
@@ -757,8 +685,8 @@ class AccountEdiCommon(models.AbstractModel):
         ) if state_code and country else self.env['res.country.state']
         if not partner and name and vat:
             partner_vals = {'name': name, 'email': email, 'phone': phone, 'is_company': True}
-            if peppol_eas and peppol_endpoint:
-                partner_vals.update({'peppol_eas': peppol_eas, 'peppol_endpoint': peppol_endpoint})
+            if routing_identifier:
+                partner_vals['routing_identifier'] = routing_identifier
             if additional_identifiers:
                 partner_vals['additional_identifiers'] = additional_identifiers
             partner = self.env['res.partner'].create(partner_vals)

--- a/addons/account_edi_ubl_cii/models/account_edi_ubl.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_ubl.py
@@ -3,21 +3,27 @@ import re
 
 from collections import defaultdict
 from markupsafe import Markup
-from stdnum.be import vat as be_vat
 
 from odoo import _, fields, models, Command
 from odoo.exceptions import UserError
 from odoo.fields import Domain
 from odoo.tools import formatLang, frozendict, html2plaintext, html_escape, unique
+from odoo.tools.partner_identifiers import get_tin_metadata_of_country
+
+from odoo.addons.base.models.res_partner_bank import sanitize_account_number
 from odoo.addons.account_edi_ubl_cii.models.account_edi_common import (
-    EAS_MAPPING,
     FloatFmt,
     GST_COUNTRY_CODES,
     UOM_TO_UNECE_CODE,
 )
-from odoo.addons.base.models.res_partner_bank import sanitize_account_number
 from odoo.addons.account_edi_ubl_cii.tools.ubl_20_optional_fields import PEPPOL_INVOICE_OPTIONAL_FIELDS, PEPPOL_INVOICE_OPTIONAL_LINE_FIELDS, PEPPOL_CREDIT_NOTE_OPTIONAL_FIELDS, PEPPOL_CREDIT_NOTE_OPTIONAL_LINE_FIELDS
 from odoo.addons.account_edi_ubl_cii.tools import Invoice, CreditNote, DebitNote
+from odoo.addons.account_edi_ubl_cii.tools.partner_identifiers import (
+    ISO_6523_ICD_CODELIST,
+    ISO_IDENTIFIERS_METADATA,
+    normalize_iso_identifier,
+    normalize_vat_for_ubl,
+)
 
 _logger = logging.getLogger(__name__)
 
@@ -749,20 +755,6 @@ class AccountEdiUBL(models.AbstractModel):
             'schemeID': None,
         }
 
-    def _ubl_add_party_identification_nodes_iso_6523_icd(self, vals):
-        nodes = vals['party_node']['cac:PartyIdentification']
-        partner = vals['party_vals']['partner']
-        commercial_partner = partner.commercial_partner_id
-        country_code = commercial_partner.country_code
-
-        if country_code == 'BE' and commercial_partner.company_registry:
-            nodes.append({
-                'cbc:ID': {
-                    '_text': be_vat.compact(commercial_partner.company_registry),
-                    'schemeID': '0208',
-                },
-            })
-
     def _ubl_add_party_identification_nodes(self, vals):
         vals['party_node']['cac:PartyIdentification'] = []
 
@@ -802,32 +794,25 @@ class AccountEdiUBL(models.AbstractModel):
         nodes = vals['party_node']['cac:PartyTaxScheme']
         partner = vals['party_vals']['partner']
         commercial_partner = partner.commercial_partner_id
-        country_code = commercial_partner.country_code
-        if not country_code:
+
+        identifier_vals = commercial_partner._get_preferred_tax_identifier_vals()
+        if not identifier_vals:
             return
+        if identifier_vals.get('category') == 'GST':
+            tax_scheme_id = 'GST'
+        else:
+            tax_scheme_id = 'VAT'
+        if identifier_vals.get('scheme') in ISO_IDENTIFIERS_METADATA:
+            normalized_value = normalize_iso_identifier(identifier_vals['scheme'], identifier_vals['value'])
+        else:
+            normalized_value = normalize_vat_for_ubl(commercial_partner.country_code, identifier_vals['value'])
 
-        if commercial_partner.vat and commercial_partner.vat != '/':
-            vat = commercial_partner.vat
-            country_code = commercial_partner.country_id.code
-            if country_code in GST_COUNTRY_CODES:
-                tax_scheme_id = 'GST'
-            else:
-                tax_scheme_id = 'VAT'
-
-            if country_code == 'HU' and not vat.upper().startswith('HU'):
-                vat = 'HU' + vat[:8]
-            if country_code == 'DK' and not vat.upper().startswith('DK'):
-                vat = f'DK{vat}'
-            if country_code == 'NO' and not vat.upper().startswith('NO'):
-                vat = f'NO{vat}'
-            if country_code == 'NO' and vat[-3:] != 'MVA':
-                vat += 'MVA'
-            nodes.append({
-                'cbc:CompanyID': {'_text': vat},
-                'cac:TaxScheme': {
-                    'cbc:ID': {'_text': tax_scheme_id},
-                },
-            })
+        nodes.append({
+            'cbc:CompanyID': {'_text': normalized_value},
+            'cac:TaxScheme': {
+                'cbc:ID': {'_text': tax_scheme_id},
+            },
+        })
 
     def _ubl_add_party_tax_scheme_nodes(self, vals):
         vals['party_node']['cac:PartyTaxScheme'] = []
@@ -836,123 +821,16 @@ class AccountEdiUBL(models.AbstractModel):
         nodes = vals['party_node']['cac:PartyLegalEntity']
         partner = vals['party_vals']['partner']
         commercial_partner = partner.commercial_partner_id
-        vat = commercial_partner.vat != '/' and commercial_partner.vat
 
-        if commercial_partner.peppol_eas in ('0106', '0190'):
-            nl_id = commercial_partner.peppol_endpoint
-        else:
-            nl_id = commercial_partner.company_registry
-
-        if commercial_partner.country_code == 'NL' and nl_id:
-            # For NL, VAT can be used as a Peppol endpoint, but KVK/OIN has to be used as PartyLegalEntity/CompanyID
-            # To implement a workaround on stable, company_registry field is used without recording whether
-            # the number is a KVK or OIN, and the length of the number (8 = KVK, 20 = OIN) is used to determine the type
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': nl_id,
-                    'schemeID': '0190' if len(nl_id) == 20 else '0106',
-                },
-            })
-        elif commercial_partner.country_code == 'LU' and commercial_partner.company_registry:
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': commercial_partner.company_registry,
-                    'schemeID': None,
-                },
-            })
-        elif commercial_partner.country_code == 'SE' and commercial_partner.company_registry:
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': ''.join(char for char in commercial_partner.company_registry if char.isdigit()),
-                },
-            })
-        elif commercial_partner.country_code == 'BE' and commercial_partner.company_registry:
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': be_vat.compact(commercial_partner.company_registry),
-                    'schemeID': '0208',
-                },
-            })
-        elif (
-            commercial_partner.country_code == 'DK'
-            and commercial_partner.peppol_eas == '0184'
-            and commercial_partner.peppol_endpoint
-        ):
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': commercial_partner.peppol_endpoint,
-                    'schemeID': '0184',
-                },
-            })
-        elif commercial_partner.country_code == 'AU' and vat:
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': commercial_partner.vat,
-                    'schemeID': '0151',
-                },
-            })
-        elif commercial_partner.country_code == 'NZ' and vat:
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': commercial_partner.vat,
-                    'schemeID': '0088',
-                },
-            })
-        elif (
-            commercial_partner.country_code == 'NO'
-            and (
-                no_id := (
-                    (
-                        'l10n_no_bronnoysund_number' in self.env['res.partner']._fields
-                        and commercial_partner.l10n_no_bronnoysund_number
-                    )
-                    or commercial_partner.company_registry
-                )
-            )
-        ):
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': no_id,
-                    'schemeID': '0192',
-                },
-            })
-        elif commercial_partner.country_code == 'NO' and commercial_partner.vat != '/':
-            if not vat.startswith('NO'):
-                vat = f'NO{vat}'
-            if not vat.endswith('MVA'):
-                vat += 'MVA'
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': vat,
-                    'schemeID': None,
-                },
-            })
-
-        elif commercial_partner.vat and commercial_partner.vat != '/':
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': commercial_partner.vat,
-                    'schemeID': None,
-                },
-            })
-        elif commercial_partner.peppol_endpoint:
-            nodes.append({
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': commercial_partner.peppol_endpoint,
-                    'schemeID': None,
-                },
-            })
+        identifier_vals = commercial_partner._get_preferred_legal_entity_identifier_vals()
+        scheme, value = (identifier_vals.get('scheme'), identifier_vals['value']) if identifier_vals else (None, None)
+        nodes.append({
+            'cbc:RegistrationName': {'_text': commercial_partner.name},
+            'cbc:CompanyID': {
+                '_text': normalize_iso_identifier(scheme, value) if scheme and value else value,
+                'schemeID': scheme if scheme in ISO_6523_ICD_CODELIST else None,
+            },
+        })
 
     def _ubl_add_party_legal_entity_nodes(self, vals):
         vals['party_node']['cac:PartyLegalEntity'] = []
@@ -2143,24 +2021,26 @@ class AccountEdiUBL(models.AbstractModel):
                     customer_values[key] = node.text
                     break
 
-        # Peppol EAS/Endpoint.
+        # Routing endpoint (cbc:EndpointID schemeID="..."): feeds both routing_identifier
+        # and additional_identifiers.
         if (node := party_node.find(".//{*}EndpointID")) is not None and node.text:
-            customer_values['peppol_endpoint'] = node.text.strip()
-            if peppol_eas := node.attrib.get('schemeID'):
-                customer_values['peppol_eas'] = peppol_eas
+            scheme, value = node.attrib.get('schemeID'), node.text.strip()
+            customer_values['routing_scheme'] = scheme
+            customer_values['routing_endpoint'] = value
+            if (meta := ISO_IDENTIFIERS_METADATA.get(scheme)):
+                customer_values.setdefault('additional_identifiers', {})[meta['key']] = value
 
         if not customer_values['vat'] and (country_code := customer_values.get('country_code')):
-            for scheme_id, field in EAS_MAPPING.get(country_code, {}).items():
-                if field == 'vat' and (vat := party_node.findtext(f".//{{*}}PartyIdentification/{{*}}ID[@schemeID='{scheme_id}']")):
-                    customer_values['vat'] = vat
-                    break
+            vat_scheme = get_tin_metadata_of_country(country_code).get('scheme')
+            if vat := party_node.findtext(f".//{{*}}PartyIdentification/{{*}}ID[@schemeID='{vat_scheme}']"):
+                customer_values['vat'] = vat
 
     def _import_ubl_retrieve_customer_search_plan(self, collected_values):
         ResPartner = self.env['res.partner']
         return [
             ResPartner._import_retrieve_customer_from_vat,
             ResPartner._import_retrieve_customer_from_additional_identifiers,
-            ResPartner._import_retrieve_customer_from_eas_endpoint,
+            ResPartner._import_retrieve_customer_from_routing_identifier,
             ResPartner._import_retrieve_customer_from_bank_account_number,
             ResPartner._import_retrieve_customer_from_email,
             ResPartner._import_retrieve_customer_from_phone,
@@ -2198,17 +2078,18 @@ class AccountEdiUBL(models.AbstractModel):
         partner_create_values = {
             'is_company': True,
         }
-        for key in ('phone', 'name', 'email', 'street', 'street2', 'zip', 'city', 'additional_identifiers'):
+        for key in ('phone', 'name', 'email', 'street', 'street2', 'zip', 'city', 'additional_identifiers', 'routing_scheme', 'routing_endpoint'):
             if value := customer_values.get(key):
                 partner_create_values[key] = value
 
-        if (peppol_eas := customer_values.get('peppol_eas')) and (peppol_endpoint := customer_values.get('peppol_endpoint')):
-            partner_create_values['peppol_eas'] = peppol_eas
-            partner_create_values['peppol_endpoint'] = peppol_endpoint
+        if (routing_scheme := customer_values.get('routing_scheme')) and (routing_endpoint := customer_values.get('routing_endpoint')):
+            partner_create_values['routing_scheme'] = routing_scheme
+            partner_create_values['routing_endpoint'] = routing_endpoint
 
         country = self._import_ubl_get_country(collected_values)
         if country:
             partner_create_values['country_id'] = country.id
+
         if vat := customer_values.get('vat'):
             partner_create_values['vat'], _country_code = self.env['res.partner']._run_vat_checks(country, vat, validation='setnull')
         return partner_create_values

--- a/addons/account_edi_ubl_cii/models/account_edi_ubl_pint.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_ubl_pint.py
@@ -2,6 +2,7 @@ from odoo import _, models
 from odoo.tools import formatLang, html2plaintext
 from odoo.tools.misc import NON_BREAKING_SPACE
 from odoo.addons.account_edi_ubl_cii.models.account_edi_common import FloatFmt
+from odoo.addons.account_edi_ubl_cii.tools.partner_identifiers import normalize_iso_identifier
 
 
 class AccountEdiUBLPint(models.AbstractModel):
@@ -147,20 +148,28 @@ class AccountEdiUBLPint(models.AbstractModel):
     def _ubl_add_party_endpoint_id_node(self, vals):
         super()._ubl_add_party_endpoint_id_node(vals)
         partner = vals['party_vals']['partner']
-        commercial_partner = partner.commercial_partner_id
-        if commercial_partner.peppol_endpoint and commercial_partner.peppol_eas:
-            vals['party_node']['cbc:EndpointID']['_text'] = commercial_partner.peppol_endpoint
-            vals['party_node']['cbc:EndpointID']['schemeID'] = commercial_partner.peppol_eas
+        identifier_vals = partner._get_preferred_routing_identifier_vals()
+        if identifier_vals:
+            normalized_value = normalize_iso_identifier(identifier_vals['scheme'], identifier_vals['value'])
+            vals['party_node']['cbc:EndpointID']['_text'] = normalized_value
+            vals['party_node']['cbc:EndpointID']['schemeID'] = identifier_vals['scheme']
 
     def _ubl_add_party_identification_nodes(self, vals):
         super()._ubl_add_party_identification_nodes(vals)
-        self._ubl_add_party_identification_nodes_iso_6523_icd(vals)
 
         nodes = vals['party_node']['cac:PartyIdentification']
         partner = vals['party_vals']['partner']
         commercial_partner = partner.commercial_partner_id
         country_code = commercial_partner.country_code
 
+        if country_code == 'BE' and (be_en := commercial_partner._get_additional_identifier('BE_EN')):
+            nodes.append({
+                'cbc:ID': {
+                    '_text': be_en,
+                    'schemeID': '0208',
+                },
+            })
+
         if not nodes and commercial_partner.ref and country_code != 'DK':  # DK-R-013
             nodes.append({
                 'cbc:ID': {
@@ -179,12 +188,12 @@ class AccountEdiUBLPint(models.AbstractModel):
         partner = vals['party_vals']['partner']
         commercial_partner = partner.commercial_partner_id
         nodes = vals['party_node']['cac:PartyTaxScheme']
-        if not nodes and commercial_partner.peppol_endpoint and commercial_partner.peppol_eas:
-            # TaxScheme based on partner's EAS/Endpoint.
+        if not nodes and commercial_partner.routing_scheme and commercial_partner.routing_endpoint:
+            # TaxScheme based on partner's Scheme/Endpoint.
             nodes.append({
-                'cbc:CompanyID': {'_text': commercial_partner.peppol_endpoint},
+                'cbc:CompanyID': {'_text': commercial_partner.routing_endpoint},
                 'cac:TaxScheme': {
-                    'cbc:ID': {'_text': commercial_partner.peppol_eas},
+                    'cbc:ID': {'_text': commercial_partner.routing_scheme},
                 },
             })
 

--- a/addons/account_edi_ubl_cii/models/account_edi_ubl_pint_eu.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_ubl_pint_eu.py
@@ -99,8 +99,8 @@ class AccountEdiUBLPintEU(models.AbstractModel):
 
             # [NL-R-003] For suppliers in the Netherlands, the legal entity identifier MUST be either a
             # KVK or OIN number (schemeID 0106 or 0190)
-            if all(node['cbc:CompanyID'] not in ('0106', '0190') for node in document_node['cac:AccountingSupplierParty']['cac:Party']['cac:PartyLegalEntity']):
-                constraints['nl_r_003'] = self._check_required_fields(vals['supplier'], 'vat')
+            if all((node.get('cbc:CompanyID') or {}).get('schemeID') not in ('0106', '0190') for node in document_node['cac:AccountingSupplierParty']['cac:Party']['cac:PartyLegalEntity']):
+                constraints['nl_r_003'] = _("%s should have a KVK or OIN number set.", vals['supplier'].display_name)
 
             # [NL-R-007] For suppliers in the Netherlands, the supplier MUST provide a means of payment
             # (cac:PaymentMeans) if the payment is from customer to supplier
@@ -119,8 +119,8 @@ class AccountEdiUBLPintEU(models.AbstractModel):
 
                 # [NL-R-005] For suppliers in the Netherlands, if the customer is in the Netherlands,
                 # the customer's legal entity identifier MUST be either a KVK or OIN number (schemeID 0106 or 0190)
-                if all(node['cbc:CompanyID'] not in ('0106', '0190') for node in document_node['cac:AccountingCustomerParty']['cac:Party']['cac:PartyLegalEntity']):
-                    constraints['nl_r_005'] = self._check_required_fields(vals['customer'], 'vat')
+                if all((node.get('cbc:CompanyID') or {}).get('schemeID') not in ('0106', '0190') for node in document_node['cac:AccountingCustomerParty']['cac:Party']['cac:PartyLegalEntity']):
+                    constraints['nl_r_005'] = _("%s should have a KVK or OIN number set.", vals['customer'].display_name)
 
         if (
             self._is_document(vals, 'credit_note')

--- a/addons/account_edi_ubl_cii/models/account_edi_xml_cii_facturx.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_xml_cii_facturx.py
@@ -178,8 +178,10 @@ class AccountEdiXmlCii(models.AbstractModel):
             'scheduled_delivery_time': self._get_scheduled_delivery_time(invoice),
             'intracom_delivery': False,
             'ExchangedDocument_vals': self._get_exchanged_document_vals(invoice),
-            'seller_specified_legal_organization': invoice.company_id.company_registry,
-            'buyer_specified_legal_organization': invoice.commercial_partner_id.company_registry,
+            # FIXME cii_22_templates hard-codes SpecifiedLegalOrganization/ID with schemeID="0002"
+            # (FR SIRENE), so only emit it when the party is French and has a SIREN/SIRET.
+            'seller_specified_legal_organization': invoice.company_id.partner_id._get_all_identifiers(enrich=True).get('FR_SIREN'),  # will retrieve SIREN from SIRET if available
+            'buyer_specified_legal_organization': invoice.commercial_partner_id._get_all_identifiers(enrich=True).get('FR_SIREN'),  # will retrieve SIREN from SIRET if available
             'ship_to_trade_party': invoice.partner_shipping_id if 'partner_shipping_id' in invoice._fields and invoice.partner_shipping_id
                 else invoice.commercial_partner_id,
             # Chorus Pro fields

--- a/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py
@@ -4,7 +4,9 @@ from odoo import _, models, Command
 from odoo.tools import html2plaintext
 from odoo.tools.float_utils import float_is_zero, float_round
 from odoo.addons.account.tools import dict_to_xml
-from odoo.addons.account_edi_ubl_cii.models.account_edi_common import EAS_MAPPING, FloatFmt
+from odoo.tools.partner_identifiers import get_tin_metadata_of_country
+
+from odoo.addons.account_edi_ubl_cii.models.account_edi_common import FloatFmt
 from odoo.addons.account_edi_ubl_cii.tools import Invoice, CreditNote, DebitNote
 from odoo.addons.account_edi_ubl_cii.tools.ubl_20_optional_fields import PEPPOL_INVOICE_OPTIONAL_FIELDS, PEPPOL_INVOICE_OPTIONAL_LINE_FIELDS, PEPPOL_CREDIT_NOTE_OPTIONAL_FIELDS, PEPPOL_CREDIT_NOTE_OPTIONAL_LINE_FIELDS
 
@@ -1115,10 +1117,9 @@ class AccountEdiXmlUBL20(models.AbstractModel):
         """ Returns a dict of values that will be used to retrieve the partner """
         vat = self._find_value(f'.//cac:{role}Party//cbc:CompanyID[string-length(text()) > 5]', tree)
         country_code = self._find_value(f'.//cac:{role}Party//cac:PostalAddress/cac:Country/cbc:IdentificationCode', tree)
-        if not vat and country_code:
-            for scheme_id, field in EAS_MAPPING.get(country_code, {}).items():
-                if field == 'vat' and (vat := self._find_value(f".//cac:{role}Party//cac:PartyIdentification/cbc:ID[@schemeID='{scheme_id}']", tree)):
-                    break
+        if not vat and country_code and (vat_scheme := get_tin_metadata_of_country(country_code).get('scheme')):
+            vat = self._find_value(f".//cac:{role}Party//cac:PartyIdentification/cbc:ID[@schemeID='{vat_scheme}']", tree)
+
         return {
             'vat': vat,
             'phone': self._find_value(f'.//cac:{role}Party//cac:Contact/cbc:Telephone', tree),

--- a/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_a_nz.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_a_nz.py
@@ -57,58 +57,3 @@ class AccountEdiXmlUbl_A_Nz(models.AbstractModel):
         # EXTENDS account.edi.xml.ubl_bis3
         super()._ubl_add_customization_id_node(vals)
         vals['document_node']['cbc:CustomizationID']['_text'] = 'urn:cen.eu:en16931:2017#conformant#urn:fdc:peppol.eu:2017:poacc:billing:international:aunz:3.0'
-
-    def _ubl_add_party_endpoint_id_node(self, vals):
-        # EXTENDS
-        super()._ubl_add_party_endpoint_id_node(vals)
-        partner = vals['party_vals']['partner']
-        commercial_partner = partner.commercial_partner_id
-
-        if commercial_partner.country_code == 'AU' and commercial_partner.vat:
-            vat = commercial_partner.vat.replace(" ", "")
-            vals['party_node']['cbc:EndpointID']['_text'] = vat
-        elif commercial_partner.country_code == 'NZ' and commercial_partner.company_registry:
-            vals['party_node']['cbc:EndpointID']['_text'] = commercial_partner.company_registry
-
-    def _ubl_add_party_tax_scheme_nodes(self, vals):
-        # EXTENDS
-        super()._ubl_add_party_tax_scheme_nodes(vals)
-        partner = vals['party_vals']['partner']
-        commercial_partner = partner.commercial_partner_id
-
-        if (
-            (commercial_partner.country_code == 'AU' and commercial_partner.vat)
-            or (commercial_partner.country_code == 'NZ' and commercial_partner.company_registry)
-        ):
-            vals['party_node']['cac:PartyTaxScheme'] = [{
-                'cbc:CompanyID': {
-                    '_text': vals['party_node']['cbc:EndpointID']['_text'],
-                    'schemeID': None,
-                },
-                'cac:TaxScheme': {
-                    'cbc:ID': {'_text': 'GST'},
-                },
-            }]
-
-    def _ubl_add_party_legal_entity_nodes(self, vals):
-        # EXTENDS
-        super()._ubl_add_party_legal_entity_nodes(vals)
-        partner = vals['party_vals']['partner']
-        commercial_partner = partner.commercial_partner_id
-
-        if commercial_partner.country_code == 'AU' and commercial_partner.vat:
-            vals['party_node']['cac:PartyLegalEntity'] = [{
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': vals['party_node']['cbc:EndpointID']['_text'],
-                    'schemeID': '0151',
-                },
-            }]
-        elif commercial_partner.country_code == 'NZ' and commercial_partner.company_registry:
-            vals['party_node']['cac:PartyLegalEntity'] = [{
-                'cbc:RegistrationName': {'_text': commercial_partner.name},
-                'cbc:CompanyID': {
-                    '_text': vals['party_node']['cbc:EndpointID']['_text'],
-                    'schemeID': '0088',
-                },
-            }]

--- a/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_bis3.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_bis3.py
@@ -1,11 +1,15 @@
 from markupsafe import Markup
 from typing import Literal
+from stdnum.no import mva
+from stdnum.be import vat as be_vat
 
 from odoo import _, api, models
+from odoo.tools.partner_identifiers import validate_identifier
 
-from stdnum.no import mva
-from stdnum.be import vat as be_vat
-CHORUS_PRO_PEPPOL_ID = "0009:11000201100044"
+from odoo.addons.account_edi_ubl_cii.models.account_edi_xml_ubl_20 import UBL_NAMESPACES
+from odoo.addons.account_edi_ubl_cii.tools.partner_identifiers import ISO_IDENTIFIERS_METADATA
+
+CHORUS_PRO_SIRET = "11000201100044"
 
 
 class AccountEdiXmlUBLBIS3(models.AbstractModel):
@@ -32,7 +36,7 @@ class AccountEdiXmlUBLBIS3(models.AbstractModel):
 
     @api.model
     def _is_customer_behind_chorus_pro(self, customer):
-        return customer.peppol_eas and customer.peppol_endpoint and f"{customer.peppol_eas}:{customer.peppol_endpoint}" == CHORUS_PRO_PEPPOL_ID
+        return customer.routing_scheme and customer.routing_endpoint and f"{customer.routing_scheme}:{customer.routing_endpoint}" == f'0009:{CHORUS_PRO_SIRET}'
 
     # -------------------------------------------------------------------------
     # EXPORT
@@ -379,19 +383,42 @@ class AccountEdiXmlUBLBIS3(models.AbstractModel):
                 ) if not mva.is_valid(vat) or len(vat) != 14 or vat[:2] != 'NO' or vat[-3:] != 'MVA' else "",
             })
 
-        if vals['supplier'].country_id.code == 'BE' and vals['supplier'].company_registry:
-            if not be_vat.is_valid(vals['supplier'].company_registry):
+        if vals['supplier'].country_id.code == 'BE' and (be_en := vals['supplier']._get_additional_identifier('BE_EN')):
+            if not be_vat.is_valid(be_en):
                 constraints.update({
-                    'PEPPOL-COMMON-R043_supplier': _('%s should have a valid KBO/BCE number in the Company ID field', vals['supplier'].display_name),
+                    'PEPPOL-COMMON-R043_supplier': _('%s should have a valid KBO/BCE number set as additional identifier (BE_EN).', vals['supplier'].display_name),
                 })
 
-        if vals['customer'].country_id.code == 'BE' and vals['customer'].company_registry:
-            if not be_vat.is_valid(vals['customer'].company_registry):
+        if vals['customer'].country_id.code == 'BE' and (be_en := vals['customer']._get_additional_identifier('BE_EN')):
+            if not be_vat.is_valid(be_en):
                 constraints.update({
-                    'PEPPOL-COMMON-R043_customer': _('%s should have a valid KBO/BCE number in the Company ID field', vals['customer'].display_name),
+                    'PEPPOL-COMMON-R043_customer': _('%s should have a valid KBO/BCE number set as additional identifier (BE_EN).', vals['customer'].display_name),
                 })
         return constraints
 
+    # -------------------------------------------------------------------------
+    # IMPORT
+    # -------------------------------------------------------------------------
+
+    def _import_retrieve_partner_vals(self, tree, role):
+        # EXTENDS account.edi.xml.ubl_20
+        partner_vals = super()._import_retrieve_partner_vals(tree, role)
+        endpoint_node = tree.find(f'.//cac:{role}Party/cac:Party/cbc:EndpointID', UBL_NAMESPACES)
+        if endpoint_node is not None:
+            scheme = endpoint_node.attrib.get('schemeID')
+            value = endpoint_node.text
+            if scheme and value:
+                meta = ISO_IDENTIFIERS_METADATA.get(scheme)
+                # Skip an endpoint whose value is malformed.
+                if not meta or validate_identifier(meta['key'], value)['valid']:
+                    partner_vals['routing_identifier'] = f"{scheme}:{value}"
+                    if meta:
+                        partner_vals.setdefault('additional_identifiers', {})[meta['key']] = value
+                elif partner_vals.get('vat') == value:
+                    # VAT is malformed: clear that bogus vat
+                    partner_vals['vat'] = None
+        return partner_vals
+
     # -------------------------------------------------------------------------
     # Sale/Purchase Order: Import
     # -------------------------------------------------------------------------

--- a/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_nlcius.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_nlcius.py
@@ -80,12 +80,11 @@ class AccountEdiXmlUbl_Nl(models.AbstractModel):
         super()._ubl_add_accounting_supplier_party_identification_nodes(vals)
         nodes = vals['party_node']['cac:PartyIdentification']
         partner = vals['party_vals']['partner']
-        commercial_partner = partner.commercial_partner_id
-
-        if commercial_partner.peppol_endpoint:
+        identifier_vals = partner._get_preferred_routing_identifier_vals()
+        if identifier_vals:
             nodes.append({
                 'cbc:ID': {
-                    '_text': commercial_partner.peppol_endpoint,
+                    '_text': identifier_vals['value'],
                     'schemeID': None,
                 },
             })
@@ -96,16 +95,24 @@ class AccountEdiXmlUbl_Nl(models.AbstractModel):
         partner = vals['party_vals']['partner']
         commercial_partner = partner.commercial_partner_id
 
-        if (
-            commercial_partner.country_code == 'NL'
-            and commercial_partner.peppol_endpoint
-        ):
+        if commercial_partner.country_code != 'NL':
+            return
+
+        # NL customer: prefer the OIN/KVK as scheme-typed PartyIdentification.
+        if oin := commercial_partner._get_additional_identifier('NL_OIN'):
             vals['party_node']['cac:PartyIdentification'] = [{
-                'cbc:ID': {
-                    '_text': commercial_partner.peppol_endpoint,
-                    'schemeID': commercial_partner.peppol_eas if commercial_partner.peppol_eas in ('0106', '0190') else None,
-                },
+                'cbc:ID': {'_text': oin, 'schemeID': '0190'},
+            }]
+        elif kvk := commercial_partner._get_additional_identifier('NL_KVK'):
+            vals['party_node']['cac:PartyIdentification'] = [{
+                'cbc:ID': {'_text': kvk, 'schemeID': '0106'},
             }]
+        else:
+            identifier_vals = commercial_partner._get_preferred_routing_identifier_vals()
+            if identifier_vals:
+                vals['party_node']['cac:PartyIdentification'] = [{
+                    'cbc:ID': {'_text': identifier_vals['value'], 'schemeID': identifier_vals['scheme']},
+                }]
 
     def _ubl_add_customization_id_node(self, vals):
         # EXTENDS account.edi.xml.ubl_bis3

--- a/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_xrechnung.py
+++ b/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_xrechnung.py
@@ -62,8 +62,9 @@ class AccountEdiXmlUbl_De(models.AbstractModel):
         node = vals['document_node']['cbc:BuyerReference']
 
         customer = vals['customer']
-        if customer.peppol_eas == "0204":
-            node['_text'] = customer.peppol_endpoint
+        if customer.routing_scheme == "0204":
+            # For B2G transactions in Germany: set the buyer_reference to the Leitweg-ID
+            node['_text'] = customer.routing_endpoint
 
         if not node['_text']:
             node['_text'] = 'N/A'
@@ -86,16 +87,15 @@ class AccountEdiXmlUbl_De(models.AbstractModel):
         partner = vals['party_vals']['partner']
         commercial_partner = partner.commercial_partner_id
 
-        if (
-            not nodes
-            and commercial_partner.peppol_eas
-        ):
-            nodes.append({
-                'cbc:CompanyID': {'_text': None},
-                'cac:TaxScheme': {
-                    'cbc:ID': {'_text': commercial_partner.peppol_eas},
-                },
-            })
+        if not nodes:
+            identifier_vals = commercial_partner._get_preferred_routing_identifier_vals()
+            if identifier_vals:
+                nodes.append({
+                    'cbc:CompanyID': {'_text': None},
+                    'cac:TaxScheme': {
+                        'cbc:ID': {'_text': identifier_vals['scheme']},
+                    },
+                })
 
     def _ubl_add_party_legal_entity_nodes(self, vals):
         # EXTENDS account.edi.xml.ubl_bis3

--- a/addons/account_edi_ubl_cii/models/account_move_send.py
+++ b/addons/account_edi_ubl_cii/models/account_move_send.py
@@ -38,7 +38,7 @@ class AccountMoveSend(models.AbstractModel):
         peppol_formats = set(self.env['res.partner']._get_peppol_formats())
         if peppol_format_moves := moves.filtered(lambda m: moves_data[m]['invoice_edi_format'] in peppol_formats):
             not_configured_company_partners = peppol_format_moves.company_id.partner_id.filtered(
-                lambda partner: not (partner.peppol_eas and partner.peppol_endpoint)
+                lambda partner: not partner.routing_identifier
             )
             if not_configured_company_partners:
                 alerts['account_edi_ubl_cii_configure_company'] = {
@@ -48,7 +48,7 @@ class AccountMoveSend(models.AbstractModel):
                     'action': not_configured_company_partners._get_records_action(),
                 }
             not_configured_partners = peppol_format_moves.partner_id.commercial_partner_id.filtered(
-                lambda partner: not (partner.peppol_eas and partner.peppol_endpoint)
+                lambda partner: not partner.routing_identifier
             )
             if not_configured_partners:
                 alerts['account_edi_ubl_cii_configure_partner'] = {
@@ -188,15 +188,17 @@ class AccountMoveSend(models.AbstractModel):
         writer = OdooPdfFileWriter()
         writer.clone_reader_document_root(reader)
 
+        # Pure Leitweg-ID partners (DE B2G) use xrechnung, not zugferd/facturx.
+        is_leitweg = bool(invoice.commercial_partner_id._get_additional_identifier('DE_LTW'))
         attachment_name = 'factur-x.xml'
-        if invoice.commercial_partner_id.country_code == 'DE' and invoice.commercial_partner_id.peppol_eas != '0204':
+        if invoice.commercial_partner_id.country_code == 'DE' and not is_leitweg:
             attachment_name = 'zugferd-invoice.xml'
 
         writer.add_attachment(attachment_name, xml_facturx, subtype='text/xml')
 
         # PDF-A.
         if ((invoice_data.get('ubl_cii_xml_options', {}).get('ubl_cii_format') in ('facturx', 'zugferd')
-                or (invoice.commercial_partner_id.country_code in ('FR', 'DE') and invoice.commercial_partner_id.peppol_eas != '0204'))
+                or (invoice.commercial_partner_id.country_code in ('FR', 'DE') and not is_leitweg))
                 and invoice.country_code in ('FR', 'DE')
                 and not writer.is_pdfa
             ):

--- a/None
+++ b/addons/account_edi_ubl_cii/models/res_company.py
@@ -0,0 +1,9 @@
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    routing_scheme = fields.Selection(related='partner_id.routing_scheme', readonly=False)
+    routing_endpoint = fields.Char(related='partner_id.routing_endpoint', readonly=False)
+    routing_identifier = fields.Char(related='partner_id.routing_identifier', readonly=False)

--- a/addons/account_edi_ubl_cii/models/res_partner.py
+++ b/addons/account_edi_ubl_cii/models/res_partner.py
@@ -1,28 +1,19 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
-import re
-from stdnum.fr import siret
+from odoo import models, fields, api
+from odoo.exceptions import (
+    UserError,
+    ValidationError,
+)
+from odoo.tools.partner_identifiers import (
+    pick_preferred_identifier,
+    validation_error_message,
+)
 
-from odoo import models, fields, api, _
-from odoo.exceptions import ValidationError
-from odoo.addons.account_edi_ubl_cii.models.account_edi_common import EAS_MAPPING
+from odoo.addons.account_edi_ubl_cii.tools.partner_identifiers import (
+    ISO_IDENTIFIERS_METADATA,
+    validate_participant_identifier,
+)
 from odoo.addons.account.models.company import PEPPOL_DEFAULT_COUNTRIES
-from odoo.addons.account.tools.partner_identifiers import ISO_IDENTIFIERS_METADATA
-from odoo.tools import single_email_re
-
-
-PEPPOL_ENDPOINT_INVALIDCHARS_RE = re.compile(r'[^a-zA-Z\d\-._~]')
-PEPPOL_ENDPOINT_INVALID_CHARS_RE_BY_EAS = {
-    '0208': re.compile(r'[^0-9]'),
-    '9925': re.compile(r'[^beBE0-9]'),
-    'EM': re.compile(r'[^a-zA-Z\d\-._@]'),
-}
-
-
-def sanitize_peppol_endpoint(peppol_endpoint, eas=None):
-    if not peppol_endpoint:
-        return peppol_endpoint
-    sanitizer = PEPPOL_ENDPOINT_INVALID_CHARS_RE_BY_EAS.get(eas, PEPPOL_ENDPOINT_INVALIDCHARS_RE)
-    return sanitizer.sub('', peppol_endpoint)
 
 
 class ResPartner(models.Model):
@@ -40,20 +31,9 @@ class ResPartner(models.Model):
         ],
     )
     is_ubl_format = fields.Boolean(compute='_compute_is_ubl_format')
-    is_peppol_edi_format = fields.Boolean(compute='_compute_is_peppol_edi_format')  # TODO remove in master
-    peppol_endpoint = fields.Char(
-        string="Peppol Endpoint",
-        help="Unique identifier used by the BIS Billing 3.0 and its derivatives, also known as 'Endpoint ID'.",
-        compute="_compute_peppol_endpoint",
-        store=True,
-        readonly=False,
-        tracking=True,
-    )
-    peppol_eas = fields.Selection(
-        string="Peppol ID",
-        help="""Code used to identify the Endpoint for BIS Billing 3.0 and its derivatives.
-             List available at https://docs.peppol.eu/poacc/billing/3.0/codelist/eas/""",
-        compute="_compute_peppol_eas",
+    routing_scheme = fields.Selection(
+        string="Routing ID",
+        compute="_compute_routing_scheme_endpoint",
         store=True,
         readonly=False,
         tracking=True,
@@ -148,17 +128,107 @@ class ResPartner(models.Model):
             ('AS', "AS2 exchange"),
             ('AU', "File Transfer Protocol"),
             ('EM', "Electronic mail"),
-        ]
+        ],
     )
-    available_peppol_eas = fields.Json(compute='_compute_available_peppol_eas')
+    routing_endpoint = fields.Char(
+        string="Routing Endpoint",
+        compute="_compute_routing_scheme_endpoint",
+        store=True,
+        readonly=False,
+        tracking=True,
+    )
+    routing_identifier = fields.Char(
+        string="EDI Routing Address",
+        compute='_compute_routing_identifier',
+        inverse='_inverse_routing_identifier',
+    )
+    available_routing_schemes = fields.Json(compute='_compute_available_routing_schemes')
 
-    @api.constrains('peppol_endpoint')
-    def _check_peppol_fields(self):
+    @api.depends_context('company')
+    @api.depends('invoice_edi_format')
+    def _compute_is_ubl_format(self):
         for partner in self:
-            if partner.peppol_endpoint and partner.peppol_eas:
-                error = self._build_error_peppol_endpoint(partner.peppol_eas, partner.peppol_endpoint)
-                if error:
-                    raise ValidationError(error)
+            partner.is_ubl_format = partner.invoice_edi_format in self._get_ubl_cii_formats()
+
+    @api.depends('commercial_partner_id', 'commercial_partner_id.vat',
+                 'commercial_partner_id.additional_identifiers', 'commercial_partner_id.country_id')
+    def _compute_routing_scheme_endpoint(self):
+        for partner in self:
+            identifier_vals = partner._get_preferred_routing_identifier_vals(force_recompute=True)
+            partner.routing_scheme = identifier_vals.get('scheme') or False
+            partner.routing_endpoint = identifier_vals.get('value') or False
+
+    @api.depends('routing_scheme', 'routing_endpoint')
+    def _compute_routing_identifier(self):
+        for partner in self:
+            partner.routing_identifier = (
+                f'{partner.routing_scheme}:{partner.routing_endpoint}'
+                if partner.routing_scheme and partner.routing_endpoint
+                else False
+            )
+
+    def _inverse_routing_identifier(self):
+        for partner in self:
+            routing_identifier = partner.routing_identifier or ''
+            scheme, sep, endpoint = routing_identifier.partition(':')
+            if routing_identifier and not sep:
+                raise UserError(self.env._("Routing identifier should be in the format 'SCHEME:ENDPOINT'."))
+            if scheme and endpoint:  # validated through '_clean_routing_endpoint'.
+                partner.write({'routing_scheme': scheme, 'routing_endpoint': endpoint})
+            else:
+                partner.write({'routing_scheme': False, 'routing_endpoint': False})
+
+    @api.depends_context('company')
+    @api.depends('company_id')
+    def _compute_available_routing_schemes(self):
+        # TO OVERRIDE
+        self.available_routing_schemes = list(dict(self._fields['routing_scheme'].selection))
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            self._clean_routing_endpoint(vals)
+        return super().create(vals_list)
+
+    def write(self, vals):
+        self._clean_routing_endpoint(vals, partners=self)
+        return super().write(vals)
+
+    def _clean_routing_endpoint(self, vals, partners=None):
+        """ Pre-create/write on a `vals` dict:
+        - normalize
+        - reject malformed values (raises ValidationError)
+        Mutates `vals` in place.
+        """
+        if (scheme := vals.get('routing_scheme')) and (endpoint := vals.get('routing_endpoint')):
+            result = validate_participant_identifier(scheme, endpoint)
+            if not result['valid']:
+                raise ValidationError(validation_error_message(self.env, result['key'], result['example']))
+            vals['routing_endpoint'] = result['value']
+
+    def _get_all_identifiers(self, enrich=False):
+        # EXTENDS 'account'
+        all_identifiers = super()._get_all_identifiers(enrich)
+        if enrich and self.routing_identifier and (metadata := ISO_IDENTIFIERS_METADATA.get(self.routing_scheme)) and metadata['key'] not in all_identifiers:
+            all_identifiers[metadata['key']] = self.routing_endpoint
+        return all_identifiers
+
+    def _get_preferred_routing_identifier_vals(self, force_recompute=False):
+        """Returns a dict {'scheme': scheme, 'value': value, ...metadata} of the preferred identifier for the given partner.
+        - When `partner.routing_identifier` is set, return it.
+        - Otherwise picks the lowest-sequence identifier carrying an ISO scheme from `_get_all_identifiers(enrich=True)`.
+        - Returns empty dict when nothing routable is available.
+        """
+        self.ensure_one()
+        partner = self.commercial_partner_id
+        if not force_recompute and partner.routing_scheme and partner.routing_endpoint and partner.routing_scheme in ISO_IDENTIFIERS_METADATA:
+            return {'scheme': partner.routing_scheme, 'value': partner.routing_endpoint}
+        identifier_vals = pick_preferred_identifier(
+            partner._get_all_identifiers(enrich=True),
+            filter_func=lambda k, v, m: m.get('scheme') in ISO_IDENTIFIERS_METADATA and v,
+            sort_key=lambda k, v, m: (m.get('sequence', 100), k),
+        )
+        return identifier_vals or {}
 
     @api.model
     def _get_ubl_cii_formats(self):
@@ -202,11 +272,11 @@ class ResPartner(models.Model):
             # return the format with the smallest sequence
             if len(formats_by_country) == 1:
                 return formats_by_country[0]
-            else:
-                if self.peppol_eas == '0204':
-                    return 'xrechnung'
-                formats_info = self._get_ubl_cii_formats_info()
-                return min(formats_by_country, key=lambda e: formats_info[e].get('sequence', 100))  # we use a sequence of 100 by default
+            # Prefer xrechnung when the partner has a Leitweg-ID (B2G in Germany).
+            if 'DE_LTW' in self.commercial_partner_id._get_all_identifiers(enrich=True):
+                return 'xrechnung'
+            formats_info = self._get_ubl_cii_formats_info()
+            return min(formats_by_country, key=lambda e: formats_info[e].get('sequence', 100))  # we use a sequence of 100 by default
         return False
 
     def _get_ubl_cii_edi_format(self):
@@ -227,105 +297,6 @@ class ResPartner(models.Model):
         formats_info = self._get_ubl_cii_formats_info()
         return [format_key for format_key, format_vals in formats_info.items() if format_vals.get('on_peppol')]
 
-    def _peppol_eas_endpoint_depends(self):
-        # field dependencies of methods _compute_peppol_endpoint() and _compute_peppol_eas()
-        # because we need to extend depends in l10n modules
-        return ['country_code', 'vat', 'additional_identifiers', 'company_registry']
-
-    @api.depends_context('company')
-    @api.depends('invoice_edi_format')
-    def _compute_is_ubl_format(self):
-        for partner in self:
-            partner.is_ubl_format = partner.invoice_edi_format in self._get_ubl_cii_formats()
-
-    @api.depends_context('company')
-    @api.depends('invoice_edi_format')
-    def _compute_is_peppol_edi_format(self):
-        for partner in self:
-            partner.is_peppol_edi_format = partner.invoice_edi_format in self._get_peppol_formats()
-
-    def _get_peppol_endpoint_value(self, country_code, field, eas):
-        self.ensure_one()
-        # Field `peppol_endpoint` can be used as placeholer for custom logic (by extending this function)
-        if field == 'peppol_endpoint':
-            return None
-
-        if field == 'additional_identifiers':
-            metadata = ISO_IDENTIFIERS_METADATA.get(eas)
-            if not metadata:
-                raise NotImplementedError()
-            value = self._get_additional_identifier(metadata['key'])
-        else:
-            value = field in self._fields and self[field]
-
-        if (
-            country_code == 'BE'
-            and field == 'company_registry'
-            and not value
-            and self.vat
-        ):
-            value = self.vat
-            if value.isalnum():
-                value = value.removeprefix(country_code)
-
-        return sanitize_peppol_endpoint(value, eas)
-
-    @api.depends('peppol_eas')
-    def _compute_peppol_endpoint(self):
-        """ If the EAS changes and a valid endpoint is available, set it. Otherwise, keep the existing value."""
-        for partner in self:
-            partner.peppol_endpoint = sanitize_peppol_endpoint(partner.peppol_endpoint, partner.peppol_eas)
-            country_code = partner._deduce_country_code()
-            if country_code in EAS_MAPPING:
-                field = EAS_MAPPING[country_code].get(partner.peppol_eas)
-                value = partner._get_peppol_endpoint_value(country_code, field, partner.peppol_eas)
-                if field and value and not partner._build_error_peppol_endpoint(partner.peppol_eas, value):
-                    partner.peppol_endpoint = value
-
-    @api.depends(lambda self: self._peppol_eas_endpoint_depends())
-    def _compute_peppol_eas(self):
-        """
-        If the country_code changes, recompute the EAS only if there is a country_code, it exists in the
-        EAS_MAPPING, and the current EAS is not consistent with the new country_code.
-        """
-        for partner in self:
-            partner.peppol_eas = partner.peppol_eas
-            country_code = partner._deduce_country_code()
-            if country_code in EAS_MAPPING:
-                eas_to_field = EAS_MAPPING[country_code]
-                if partner.peppol_eas not in eas_to_field.keys():
-                    new_eas = next(iter(EAS_MAPPING[country_code].keys()))
-                    # Iterate on the possible EAS until a valid one is found
-                    for eas, field in eas_to_field.items():
-                        if field and field in partner._fields:
-                            value = partner._get_peppol_endpoint_value(country_code, field, eas)
-                            if value and not partner._build_error_peppol_endpoint(eas, value):
-                                new_eas = eas
-                                break
-                    partner.peppol_eas = new_eas
-
-    @api.depends_context('company')
-    @api.depends('company_id')
-    def _compute_available_peppol_eas(self):
-        # TO OVERRIDE
-        self.available_peppol_eas = list(dict(self._fields['peppol_eas'].selection))
-
-    def _build_error_peppol_endpoint(self, eas, endpoint):
-        """ This function contains all the rules regarding the peppol_endpoint."""
-        if eas == '0208' and not re.match(r"^\d{10}$", endpoint):
-            return _("The Peppol endpoint is not valid. The expected format is: 0239843188")
-        if eas == '0009' and not siret.is_valid(endpoint):
-            return _("The Peppol endpoint is not valid. The expected format is: 73282932000074")
-        if eas == '0007' and not re.match(r"^\d{10}$", endpoint):
-            return _("The Peppol endpoint is not valid. "
-                     "It should contain exactly 10 digits (Company Registry number)."
-                     "The expected format is: 1234567890")
-        if eas == 'EM' and not single_email_re.match(endpoint):
-            return _("The Peppol endpoint is not valid. A valid email is required")
-        invalid_chars_re = PEPPOL_ENDPOINT_INVALID_CHARS_RE_BY_EAS.get(eas, PEPPOL_ENDPOINT_INVALIDCHARS_RE)
-        if invalid_chars_re.search(endpoint) or not 1 <= len(endpoint) <= 50:
-            return _("The Peppol endpoint (%s) is not valid. It should contain only letters and digit.", endpoint)
-
     @api.model
     def _get_edi_builder(self, invoice_edi_format):
         if invoice_edi_format == 'xrechnung':
@@ -343,14 +314,14 @@ class ResPartner(models.Model):
             return self.env['account.edi.xml.ubl_sg']
 
     @api.model
-    def _import_retrieve_customer_from_eas_endpoint(self, customer_values):
-        peppol_eas = customer_values.get('peppol_eas')
-        peppol_endpoint = customer_values.get('peppol_endpoint')
-        if not peppol_eas or not peppol_endpoint:
+    def _import_retrieve_customer_from_routing_identifier(self, customer_values):
+        routing_scheme = customer_values.get('routing_scheme')
+        routing_endpoint = customer_values.get('routing_endpoint')
+        if not routing_scheme or not routing_endpoint:
             return
 
         return {
             'criteria': [{
-                'domain': [('peppol_eas', '=', peppol_eas), ('peppol_endpoint', '=', peppol_endpoint)],
+                'domain': [('routing_scheme', '=', routing_scheme), ('routing_endpoint', '=', routing_endpoint)],
             }],
         }
